packages/extension/src/background/handlers/State.ts

```diff
@@ -213,6 +222,40 @@ export default class State {
     return this.#signRequests[id];
   }
 
+  public rpcSend (request: RequestRpcSend): Promise<JsonRpcResponse> {
+    assert(this.#provider, 'Cannot call pub(rpc.subscribe) before provider has been set');
+
+    return this.#provider.send(request.method, request.params);
+  }
+
+  // Change the provider the extension is exposing
+  public rpcSetProvider (provider: RequestRpcProvider): Promise<null> {
+    switch (provider.type) {
+      case 'WsProvider': {
+        this.#provider = new WsProvider(provider.payload);
```

Comment:
haven't tested, but looking at the code it seems me to me that

- when the page closes, the wsprovider connections it has made should be closed, which doesn't seem to be the case here (subscriptions are left active), and
- unexpected behaviour if two pages use the extension (i.e. if two pages are opened which use `rpcSetProvider`, the latter will create a new WsProvider, and the existing connections from the former page will use the previous WsProvider which doesn't have any more references to it, so memory leaks, while the next subscriptions from the former page will use the new WsProvider -- also any page can change the extension provider endpoint used by all the other pages?). in my PR I had solved this by mapping a page (chrome.runtime.Port) to a provider ([code](https://github.com/polkadot-js/extension/pull/88/files#diff-08c560d3145c89a48bc10bc086ce5fe0R239)) and using one WsProvider connection per page